### ISSUE TITLE
Averaging over time with ft_freqstatistics and 'union' parameter

### ISSUE THREAD

**Issue by rtheze** [fieldtrip/fieldtrip#1089]
I have been trying to run ft_freqstatistics with ft_statfun_actvsblT as statistic. Indeed I am trying to test the activation against the baseline, but my trials have various length and my baseline is constant. Therefore I have redefined my trials into baseline and trials, and I run ft_freqstatistics with the parameter cfg.select ='union' since the two will not overlap. I also want to use the cfg.avgovertime parameter in order to reduce the time dimension, which is different for each trial. The code is as follow:

```
    cfg                  = [];
    cfg.channel          = 'all';
    cfg.latency          = 'all';
    cfg.frequency        = 'all';
    cfg.avgoverchan      = 'no';   
    cfg.avgovertime      = 'yes';            
    cfg.avgoverfreq      = 'no';                 
    cfg.parameter        = 'powspctrm';
    cfg.select           = 'union';
    cfg.nanmean          = 'yes';

    cfg.tail             = 0;                                               
    cfg.alpha            = 0.025;                                           
    cfg.numrandomization = 10;
    
    cfg.method           = 'montecarlo';
    cfg.statistic        = 'ft_statfun_actvsblT'; 
    cfg.computeprob      = 'yes';
                 
    cfg.design(1,:)      = [ones(1,size(tf_activation{1}.powspctrm,1)) 2*ones(1,size(tf_baseline{1}.powspctrm,1))];   
    cfg.design(2,:)      = [1:size(tf_activation{1}.powspctrm,1) 1:size(tf_baseline{1}.powspctrm,1)];
    cfg.ivar             = 1;
    cfg.uvar             = 2;
    
    [statV1{s}]          = ft_freqstatistics(cfg, tf_activation{1}, tf_baseline{1}); 

```
tf_activation is a struct with fields: cfg, label (257 x 1 cell), dimord (rpt_chan_freq_time), freq (1x23 double), time (1x64 double), powspctrm (58 x 257 x 23 x 64 double), cumtapcnt (58 x 23 double). 

The field powspctrm has nan for every trial that has not 64 time points. 
The struct tf_baseline is similar but with 6 time points for every trial. 

When I try to run the script with these two parameters I get an error message as such: 

> 'cfg.select ''union'' in combination with averaging across one of the dimensions is not possible'

This error message comes from ft_selectdata. But if I go trhough the script the paramter 'union' seems adequatly programmed for averaging over time. In fact, if I shut this error message, the script runs fine. 

Thus I am wondering if this error message is a vestige of an earlier version or if there is a reason for not allowing the combination of the two, and then why is that?

In advance, thank you for your insights!

**Comment by robertoostenveld** [comment]
Since (I guess) the implementation is not guaranteed to be robust for all use cases, I think it is disabled with the error. Looking at the code of ft_selectdata, I also don't see why it is disabled.

You could do them separately. I.e. first use ft_selectdata to concatenate without averaging , then use ft_selectdata to average over time, then do ft_freqstatistics.

it would be interesting to actually see whether doing them separate and doing them in one go has the same results...

**Comment by robertoostenveld** [comment]
Note that I would expect your `cumtapcnt` to be 58x1, not 58x23. Could you check?

**Comment by robertoostenveld** [comment]
I doubt that the code is actually working as you suggest. When I do the following, the outputs are identical, but don't have the content that I expect.

```
%%
% make two datasets, with zero overlap

tf_activation = [];
tf_activation.cfg = struct();
tf_activation.label = cellfun(@num2str, num2cell(1:10), 'UniformOutput', false);
tf_activation.dimord = 'rpt_chan_freq_time';
tf_activation.freq = 1:10;
tf_activation.time = 1:10;
tf_activation.powspctrm = randn(30, 10, 10, 10);
tf_activation.cumtapcnt = ones(30, 1);

tf_baseline = [];
tf_baseline.cfg = struct();
tf_baseline.label = cellfun(@num2str, num2cell(11:15), 'UniformOutput', false);
tf_baseline.dimord = 'rpt_chan_freq_time';
tf_baseline.freq = 11:15;
tf_baseline.time = 11:15;
tf_baseline.powspctrm = randn(30, 5, 5, 5);
tf_baseline.cumtapcnt = ones(30, 1);

%%
% do the averaging and concatenation in one go

cfg = [];
cfg.channel          = 'all';
cfg.latency          = 'all';
cfg.frequency        = 'all';
cfg.parameter        = 'powspctrm';
cfg.select           = 'union';

cfg.avgoverrpt       = 'yes';
cfg.avgoverchan      = 'no';
cfg.avgovertime      = 'no';
cfg.avgoverfreq      = 'no';
tf_combined1r = ft_selectdata(cfg, tf_activation, tf_baseline);

cfg.avgoverrpt       = 'no';
cfg.avgoverchan      = 'yes';
cfg.avgovertime      = 'no';
cfg.avgoverfreq      = 'no';
tf_combined1c = ft_selectdata(cfg, tf_activation, tf_baseline);

cfg.avgoverrpt       = 'no';
cfg.avgoverchan      = 'no';
cfg.avgovertime      = 'yes';
cfg.avgoverfreq      = 'no';
tf_combined1t = ft_selectdata(cfg, tf_activation, tf_baseline);

cfg.avgoverrpt       = 'no';
cfg.avgoverchan      = 'no';
cfg.avgovertime      = 'no';
cfg.avgoverfreq      = 'yes';
tf_combined1f = ft_selectdata(cfg, tf_activation, tf_baseline);

% this is without averaging, needed for comparison below
cfg.avgoverrpt       = 'no';
cfg.avgoverchan      = 'no';
cfg.avgovertime      = 'no';
cfg.avgoverfreq      = 'no';
tf_combined1 = ft_selectdata(cfg, tf_activation, tf_baseline);


%%
% do the averaging after the concatenation

cfg = [];
cfg.select           = 'union'; % intersect works fine

cfg.avgoverrpt       = 'yes';
cfg.avgoverchan      = 'no';
cfg.avgovertime      = 'no';
cfg.avgoverfreq      = 'no';
tf_combined2r = ft_selectdata(cfg, tf_combined1);
assert(isequaln(tf_combined1r.powspctrm, tf_combined2r.powspctrm))

cfg.avgoverrpt       = 'no';
cfg.avgoverchan      = 'yes';
cfg.avgovertime      = 'no';
cfg.avgoverfreq      = 'no';
tf_combined2c = ft_selectdata(cfg, tf_combined1);
assert(isequaln(tf_combined1c.powspctrm, tf_combined2c.powspctrm))

cfg.avgoverrpt       = 'no';
cfg.avgoverchan      = 'no';
cfg.avgovertime      = 'yes';
cfg.avgoverfreq      = 'no';
tf_combined2t = ft_selectdata(cfg, tf_combined1);
assert(isequaln(tf_combined1t.powspctrm, tf_combined2t.powspctrm))

cfg.avgoverrpt       = 'no';
cfg.avgoverchan      = 'no';
cfg.avgovertime      = 'no';
cfg.avgoverfreq      = 'yes';
tf_combined2f = ft_selectdata(cfg, tf_combined1);
assert(isequaln(tf_combined1f.powspctrm, tf_combined2f.powspctrm))
```

looking at it in more detail, it actually seems that `union` is not working, correctly, also not without averaging

**Comment by robertoostenveld** [comment]
sorry, please ignore the previous comments. I was mistakenly assuming that ft_selectdata was also appending. It used to do that in a previous version, but not any more.

this is working just fine:

```
timelock1 = [];
timelock1.label = {'1'};
timelock1.time = 1:10;
timelock1.avg = randn(1,10);

timelock2 = [];
timelock2.label = {'2'};
timelock2.time = 11:15;
timelock2.avg = randn(1,5);

cfg = [];
cfg.select = 'union';
[sel1, sel2] = ft_selectdata(cfg, timelock1, timelock2)
```

**Comment by rtheze** [comment]
> Note that I would expect your `cumtapcnt` to be 58x1, not 58x23. Could you check?

Nope it's 58x23, but I guess it's normal since I have repetition and frequencies dimensions? In the usual situations, it's only one of the two.

**Comment by rtheze** [comment]
> sorry, please ignore the previous comments. I was mistakenly assuming that ft_selectdata was also appending. It used to do that in a previous version, but not any more.
> 
> this is working just fine:
> 
> ```
> timelock1 = [];
> timelock1.label = {'1'};
> timelock1.time = 1:10;
> timelock1.avg = randn(1,10);
> 
> timelock2 = [];
> timelock2.label = {'2'};
> timelock2.time = 11:15;
> timelock2.avg = randn(1,5);
> 
> cfg = [];
> cfg.select = 'union';
> [sel1, sel2] = ft_selectdata(cfg, timelock1, timelock2)
> ```

'Union' was working perfectly fine when I tried it alone without averaging, it's when I tried to average over time in addition to that, that the error message showed up. 

But, to be sure I am following, you suggest to separately call ft_selectdata to first format the data before calling ft_freqstatistics to be sure that it does the same as calling the two paramter at once with ft_freqstatistics?

**Comment by robertoostenveld** [comment]
after some further testing, I am getting more confident that the error message is not needed any more.

**Comment by robertoostenveld** [comment]
You are right, the number of tapers can be different per trial AND per frequency. Hence 58x23. I noticed one of the FT functions (checkmate?) giving a warning on it, so that needs to be improved.

**Comment by rtheze** [comment]
Ok, great, thank you! So that means I can go on with my analysis?
Also should I request a pull for disabling that error message?

**Comment by robertoostenveld** [comment]
please have a look at the test script in the PR

**Comment by rtheze** [comment]
I ran the test script with my original data and I don't see a difference. It confirms it works I suppose.

**Comment by schoffelen** [comment]
Hi guys, can this issue be closed after merging the respective PR?

**Comment by robertoostenveld** [comment]
I think so... I just merged #1090 into master. let's give it a night for all dashboard test scripts to run, and if nothing breaks we can close it here as well.
